br/pkg/backup/prepare_snap/prepare.go

```diff
@@ -0,0 +1,420 @@
+	if r.maybeID == 0 {
+		return fmt.Sprintf("range%s", rng)
+	}
+	return fmt.Sprintf("region(id=%d, range=%s)", r.maybeID, rng)
+}
+
+func (r region) Less(than btree.Item) bool {
+	return bytes.Compare(r.startKey, than.(region).startKey) < 0
+}
+
+type Preparer struct {
+	/* Environments. */
+	env Env
+
+	/* Internal Status. */
+	inflightReqs         map[uint64]metapb.Region
+	failedRegions        []region
+	waitApplyDoneRegions btree.BTree
+	retryTime            int
+	nextRetry            *time.Timer
+
+	/* Internal I/O. */
+	eventChan chan event
+	clients   map[uint64]*prepareStream
+
+	/* Interface for caller. */
+	waitApplyFinished bool
+
+	/* Some configurations. They aren't thread safe.
+	   You may need to configure them before starting the Preparer. */
+	RetryBackoff  time.Duration
+	RetryLimit    int
+	LeaseDuration time.Duration
+}
+
+func New(env Env) *Preparer {
+	prep := &Preparer{
+		env: env,
+
+		inflightReqs:         make(map[uint64]metapb.Region),
+		waitApplyDoneRegions: *btree.New(16),
+		eventChan:            make(chan event, 128),
+		clients:              make(map[uint64]*prepareStream),
+
+		RetryBackoff:  defaultRetryBackoff,
+		RetryLimit:    defaultMaxRetry,
+		LeaseDuration: defaultLeaseDur,
+	}
+	return prep
+}
+
+func (p *Preparer) MarshalLogObject(om zapcore.ObjectEncoder) error {
+	om.AddInt("inflight_requests", len(p.inflightReqs))
+	for _, r := range p.inflightReqs {
+		om.AddString("simple_inflight_region", region{maybeID: r.Id, startKey: r.StartKey, endKey: r.EndKey}.String())
+	}
+	om.AddInt("failed_requests", len(p.failedRegions))
+	for _, r := range p.failedRegions {
+		om.AddString("simple_failed_region", r.String())
+	}
+	err := om.AddArray("connected_stores", zapcore.ArrayMarshalerFunc(func(ae zapcore.ArrayEncoder) error {
+		for id := range p.clients {
+			ae.AppendUint64(id)
+		}
+		return nil
+	}))
+	if err != nil {
+		return err
+	}
+	om.AddInt("retry_time", p.retryTime)
+	om.AddBool("wait_apply_finished", p.waitApplyFinished)
+	return nil
+}
+
+// DriveLoopAndWaitPrepare drives the state machine and block the
+// current goroutine until we are safe to start taking snapshot.
+//
+// After this invoked, you shouldn't share this `Preparer` with any other goroutines.
+//
+// After this the cluster will enter the land between normal and taking snapshot.
+// This state will continue even this function returns, until `Finalize` invoked.
+// Splitting, ingesting and conf changing will all be blocked.
+func (p *Preparer) DriveLoopAndWaitPrepare(ctx context.Context) error {
+	logutil.CL(ctx).Info("Start drive the loop.", zap.Duration("retry_backoff", p.RetryBackoff),
+		zap.Int("retry_limit", p.RetryLimit),
+		zap.Duration("lease_duration", p.LeaseDuration))
+	p.retryTime = 0
+	if err := p.prepareConnections(ctx); err != nil {
+		log.Error("failed to prepare connections", logutil.ShortError(err))
+		return errors.Annotate(err, "failed to prepare connections")
+	}
+	if err := p.MaybeFinish(ctx); err != nil {
```

Comment:
why finish at beginning? Will `checkHole` return success if there is no entry in BtreeMap?

Reply:
It won't, and then we will generate the requests we need which will fill these holes.